engine/core/src/main/java/org/datacleaner/util/CrosstabReducerHelper.java

```diff
@@ -0,0 +1,114 @@
+     *            - list of dimensions
+     * @param partialCrosstab
+     *            - crosstab
+     */
+    public static void createDimensions(List<CrosstabDimension> crosstabDimensions,
+            final Crosstab<Number> partialCrosstab) {
+        if (partialCrosstab != null) {
+            final List<CrosstabDimension> dimensions = partialCrosstab.getDimensions();
+            for (CrosstabDimension dimension : dimensions) {
+                if (!dimensionExits(crosstabDimensions, dimension)) {
+                    crosstabDimensions.add(dimension);
+                }
+            }
+        }
+    }
+
+    private static boolean dimensionExits(Collection<CrosstabDimension> list, CrosstabDimension dimension) {
+        if (list.size() > 0) {
+            boolean allreadyExits = false;
+            for (CrosstabDimension dim : list) {
+                if (dimension.equals(dim)) {
+                    allreadyExits = true;
+                    break;
+                }
+            }
+            return allreadyExits;
+        }
+        return false;
+    }
+
+    /**
+     * Add the values of partial crosstab to the main crosstab
+     * 
+     * @param mainCrosstab
+     *            - main crosstab
+     * @param partialCrosstab
+     *            - partial crosstab
+     */
+    public static void addData(final Crosstab<Number> mainCrosstab, final Crosstab<Number> partialCrosstab) {
+        if (partialCrosstab != null) {
+
+            final CrosstabNavigator<Number> mainNavigator = new CrosstabNavigator<Number>(mainCrosstab);
+            final CrosstabNavigator<Number> nav = new CrosstabNavigator<Number>(partialCrosstab);
+            final List<CrosstabDimension> dimensions = partialCrosstab.getDimensions();
+            for (CrosstabDimension dimension : dimensions) {
+                final List<String> categories = dimension.getCategories();
+                for (String category : categories) {
+                    final Number categoryValue = nav.where(dimension, category).safeGet(null);
+                    final CrosstabNavigator<Number> whereToPut = mainNavigator.where(dimension, category);
+                    if (categoryValue != null) {
+                        final Number oldValue = whereToPut.safeGet(null);
+                        if (oldValue != null) {
+                            final Number newValue = sum(oldValue, categoryValue);
+                            whereToPut.put(newValue);
+                        } else {
+                            whereToPut.put(categoryValue);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private static Number sum(Number n1, Number n2) {
+        if (isIntegerType(n1) && isIntegerType(n2)) {
+            return BigInteger.valueOf(n1.longValue()).add(BigInteger.valueOf(n2.longValue()));
+        } 
+        return new BigDecimal(n1.doubleValue()).add(new BigDecimal(n2.doubleValue()));
```

Comment:
actually there's still a risk of precision loss here ... what if n1 or n2 is itself a BigInteger or BigDecimal - then I guess the .longValue and .doubleValue calls will loose precision or could cause overflow. Yuck. Smells a bit like gluing code that someone else has probably done a bunch of times before.

I can live with it like it is, because I think the risk is quite small, but ideally we should create or find a utility class that does this with more guarantees that we are not loosing precision etc.